src/ui/auth/uaa/uaa.go

```diff
@@ -43,51 +38,85 @@ func GetClient() (uaa.Client, error) {
 		Endpoint:      UAASettings.Endpoint,
 		SkipTLSVerify: !UAASettings.VerifyCert,
 	}
-	client, err = uaa.NewDefaultClient(cfg)
-	return client, err
+	return uaa.NewDefaultClient(cfg)
+}
+
+// Auth is the implementation of AuthenticateHelper to access uaa for authentication.
+type Auth struct {
+	sync.Mutex
+	client uaa.Client
 }
 
-func doAuth(username, password string, client uaa.Client) (*models.User, error) {
-	t, err := client.PasswordAuth(username, password)
+//Authenticate ...
+func (u *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
+	if err := u.ensureClient(); err != nil {
+		return nil, err
+	}
+	t, err := u.client.PasswordAuth(m.Principal, m.Password)
 	if t != nil && err == nil {
 		//TODO: See if it's possible to get more information from token.
-		u := &models.User{
-			Username: username,
+		user := &models.User{
+			Username: m.Principal,
 			Password: "1234567ab",
-			Email:    username + "@placeholder.com",
-			Realname: username,
-		}
-		err = dao.OnBoardUser(u)
-		if err == nil {
-			return u, nil
+			Email:    m.Principal + "@uaa.placeholder",
+			Realname: m.Principal,
 		}
+		err = u.OnBoardUser(user)
+		return user, err
 	}
 	return nil, err
 }
 
-// Auth is the implementation of AuthenticateHelper to access uaa for authentication.
-type Auth struct{}
+// OnBoardUser will check if a user exists in user table, if not insert the user and
+// put the id in the pointer of user model, if it does exist, return the user's profile.
+func (u *Auth) OnBoardUser(user *models.User) error {
+	return dao.OnBoardUser(user)
+}
 
-//Authenticate ...
-func (u *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
-	client, err := GetClient()
+// SearchUser search user on uaa server, transform it to Harbor's user model
+func (u *Auth) SearchUser(username string) (*models.User, error) {
+	if err := u.ensureClient(); err != nil {
+		return nil, err
+	}
+	l, err := u.client.SearchUser(username)
 	if err != nil {
 		return nil, err
 	}
-	return doAuth(m.Principal, m.Password, client)
-}
-
-// OnBoardUser will check if a user exists in user table, if not insert the user and
-// put the id in the pointer of user model, if it does exist, return the user's profile.
-// func (u *Auth) OnBoardUser(user *models.User) error {
-// 	panic("not implemented")
-// }
+	if len(l) == 0 {
+		return nil, nil
+	}
+	if len(l) > 1 {
+		return nil, fmt.Errorf("Multiple entries found for username: %s", username)
+	}
+	e := l[0]
+	email := username + "@uaa.placeholder"
```

Comment:
Suggest to move them to OnBoardUser() to avoid unformatted data

Reply:
will do , thanks!